Fix moose deprecation warnings

lindsayad opened the pull request:
Closes #14047 #13909 #14018

lindsayad committed:
Fix deprecated moose modules

Refs #14047 #13909

lindsayad committed:
Fix deprecated moose test

Refs #13909

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 19e96cb6ce080554cb10910c3b49a0b3c7142403
+Subproject commit e2c3745c7ef556911350e1c698b23804a1ea96a8
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/399391/) on 13bc160 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14061/site)

This comment will be updated on new commits.

lindsayad committed:
Remove generators_no_mesh_block test

lindsayad committed:
Update documentation

lindsayad committed:
Make sure that parallel_type gets applied

lindsayad commented:
This single failure is gonna be a fun one to debug :-)

tophmatthews commented on modules/combined/test/tests/contact_verification/hertz_cyl/half_symm_q4/hertz_cyl_half_1deg_template1.i:
```diff
@@ -72,6 +73,7 @@
   [./TensorMechanics]
     use_displaced_mesh = true
     save_in = 'saved_x saved_y'
+    extra_vector_tags = 'ref'
```
This isn't in the master action yet is it?

tophmatthews replied:
I think adding
```
  params.addParam<std::vector<TagName>>("extra_vector_tags",
                                        "The extra tags for the vectors this Kernel should fill");
```
to `TensorMechanicsActionBase` input parameters should take care of this. I'm pretty sure tagging, and thus extra_vector_tags, is only applicable to kernels, BCs, and constraints, so the `applyParameters` should take care of the rest.

tophmatthews commented on modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
```diff
@@ -56,6 +56,9 @@ validParams<TensorMechanicsAction>()
   params.addParam<std::string>(
       "strain_base_name",
       "The base name used for the strain. If not provided, it will be set equal to base_name");
+  params.addParam<std::vector<TagName>>(
```
Damnit! I didn't see this...

tophmatthews replied:
It'd be nice to get this in sooner rather than later so I can start applying reference residual to your thermal contact methods.

lindsayad commented:
Just debugging one very tricky test 

> On Sep 24, 2019, at 10:06 AM, Topher Matthews <notifications@github.com> wrote:
> 
> @tophmatthews commented on this pull request.
> 
> In modules/tensor_mechanics/src/actions/TensorMechanicsAction.C:
> 
> > @@ -56,6 +56,9 @@ validParams<TensorMechanicsAction>()
>    params.addParam<std::string>(
>        "strain_base_name",
>        "The base name used for the strain. If not provided, it will be set equal to base_name");
> +  params.addParam<std::vector<TagName>>(
> It'd be nice to get this in sooner rather than later so I can start applying reference residual to your thermal contact methods.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

lindsayad commented:
It's pretty incredible that only one test fails because of this:

This could coincide with #14077...but right now when we are using `MeshGenerators` we are adding geometric relationship managers to a `MeshBase` object that is not the `MeshBase` object actually used in the simulation.

Sequence of events:

1. SetupMeshAction::act with `_current_task == "set_mesh_base"`
2. MooseApp::attachRelationshipManagers: attaches geometric relationship managers to the `MeshBase` set in 1.
3. SetupMeshAction::act with `_current_task == "init_mesh"`: this calls MooseMesh::init which then calls MeshGeneratorMesh::buildMesh which promptly proceeds to set the `MeshBase` object to that created through `MooseApp::executeMeshGenerators`. This `MeshBase` object does not have any of our custom geometric ghosting functors!

lindsayad committed:
Make sure RMs added to the correct MeshBase

Previously when using MeshGenerators, RMs were added to a
MeshBase that wasn't even used in the simulation

lindsayad committed:
Perform a meaningful check for whether MeshGenerators are present

lindsayad committed:
Dont use the MeshGeneratorMes when recovering

andrsd commented on framework/src/actions/SetupMeshAction.C:
```diff
@@ -258,7 +258,12 @@ SetupMeshAction::act()
+      if (_awh.hasActions("add_mesh_generator"))
+        _mesh->setMeshBase(_app.getMeshGeneratorMesh());
+      else
+        _mesh->setMeshBase(_mesh->buildMeshBaseObject());
```
This will create a conflict with #14077, but it is easy to fix: just need to add the guards. For now, just for the else branch. When I start using MeshGenerators in the component system, I might need to do something about if-branch, but may be not - it is hard to predict that...

lindsayad committed:
Use new hasMeshBase API

lindsayad committed:
During recover, run MeshGenerators for sub-apps

tophmatthews commented on modules/combined/test/tests/contact_verification/hertz_cyl/half_symm_q4/hertz_cyl_half_1deg_template1.i:
```diff
@@ -10,7 +10,8 @@
   solution_variables = 'disp_x disp_y'
-  reference_residual_variables = 'saved_x saved_y'
+  extra_tag_vectors = 'ref'
+  reference_vector = 'ref'
```
Almost seems like `extra_tag_vectors` and `reference_vector` can be automated. In general, `extra_tag_vectors=ref` is added to kernels and constrains, and not BCs. Is there cases where this is not true? Could `extra_tag_vectors` automatically use `ref` if `type=ReferenceResidualProblem`?

Certainly, `solution_variables` can be automated? Just thinking about trying to make `ReferenceResidualProblem` easier in future PRs.

tophmatthews replied:
@bwspenc are there cases where the old `save_in` method breaks from the kernels/constrains and not bcs pattern?

bwspenc replied:
@tophmatthews Typically we just put it in the kernels. If you save into the reference vector for the kernels and constraints, the only nonzero values in the vector would be the reactions at the nodes when the solution is converged. 

Do we actually put it in the constraints?  I don't see where that is done in this input, for example. I've been a little out of the loop on the recent changes here. I don't think we have an option in the Contact action to specify a tag vector. I don't think we ever did that with save_in either.  If we don't put it in the constraints, I guess we'd also get the contact forces in the vector, which probably isn't a bad thing. That would probably behave better in cases where you have contact but nothing applied at BCs.

The whole idea of this approach is that it's giving you a very rough measure of something to converge relative to. There's not really a strictly right or a wrong way to do it.

bwspenc replied:
@tophmatthews I definitely like the idea of streamlining this. I guess the challenge is that you have to specify the tag vector both in the Action that sets up the kernels and the ReferenceResidualProblem block.

I was hoping that eventually the stuff in ReferenceResidualProblem could just be rolled into FEProblem. It's definitely good to have it moved into the framework, but we still have the problem that if for some reason someone needs to use another Problem type, they can't use this technique, which I think should really be our default way of checking convergence.

tophmatthews replied:
Why do you have to specify the tag vector? Can't a default name be used, and the kernels check the problem type?

tophmatthews replied:
Rolled into FEProblem would be great...but my guess is that'd be a battle. I'm just happy it's finally in the framework!

lindsayad replied:
> Almost seems like `extra_tag_vectors` and `reference_vector` can be automated. In general, `extra_tag_vectors=ref` is added to kernels and constrains, and not BCs. Is there cases where this is not true? Could `extra_tag_vectors` automatically use `ref` if `type=ReferenceResidualProblem`?
> 
> Certainly, `solution_variables` can be automated? Just thinking about trying to make `ReferenceResidualProblem` easier in future PRs.

The only automation I would support is a default parameter value for `reference_vector` for the `ReferenceResidualProblem`. Here's why I don't support the others:

- `extra_tag_vectors`: this parameter is inherited from `FEProblemBase` and the user could want to create many of these, totally unrelated to convergence critera for `ReferenceResidualProblem`
- `extra_vector_tags`: this parameter is inherited from the `TaggingInterface` and again the user could want to specify many of these, not just for use for reference residuals. Moreover, other physics may prefer to specify saving into the reference residual vector from BCs and ***not*** kernels, just like in my test case. Indeed the [`ReferenceResidualProblem` documentation](https://mooseframework.org/source/problems/ReferenceResidualProblem.html) suggests that integrated fluxes can be a good choice for a reference quantity.

tophmatthews replied:
`solution_variables`? That's the one that always catches me...

tophmatthews replied:
Also, almost seems like there could be a `referance_residual_tag` alongside the others...but I get the point

lindsayad replied:
> `solution_variables`? That's the one that always catches me...

As the class stands now, this could be automated since it actually errors out if the size of `solution_variables` does not match the number of variables in the non-linear system. ***But*** I want to remove that error since the relative convergence criteria in `ReferenceResidualProblem` doesn't make sense for variables that only have residual contributions from one object, like in the case of the Lagrange Multiplier variable for mechanical or thermal contact.

lindsayad replied:
Anyways explicit specification of all these parameters should certainly be allowed. If someone wants to submit PRs providing behind-the-scenes automation if the parameters are not provided, then I would review them 😄

lindsayad replied:
BTW @tophmatthews I know you have been using `ReferenceResidualProblem` with mortar contact. I would be curious to see a solve history; the only way I can imagine this working is through an absolute tolerance...?

tophmatthews replied:
You're assuming I know what I'm doing there...not sure I'm doing it right. My Lagrange multiplier for thermal contact has a terrible norm once in mechanical contact. I use save_in for temperature and displacements, but assign nothing to the Lagrange multiplier constraint. I think that's the right way no?

lindsayad replied:
Well your reference quantity will be zero...it's pretty hard to reduce that value by 6-8 orders of magnitude 😃 So I think if you checked why you are achieving convergence, it has to be through an absolute tolerance. If you set your `nl_abs_tol = 0`, I think perhaps you would never see convergence...

tophmatthews replied:
I assumed if the reference quantity is zero, and your relative tolerance is 1e-6, then if the norm was less than 1e-6 it would be okay?

lindsayad replied:
This is the convergence check:
```c++
((_group_resid[i] < _group_ref_resid[i] * rtol) || (_group_resid[i] < abstol))
```
If `group_ref_resid` is zero, then `group_ref_resid * rtol` will also be zero, hence the LHS boolean will always evaulate to `false`.

bwspenc replied:
Couldn't we make it so that `solution_variables` is optional? If you don't specify it, it would just default to using the full set of nonlinear variables. I originally had that be specified so that the correspondence of reference variables to solution variables was defined, but that's probably not needed anymore with the tagging option.

lindsayad replied:
In the parameter sense it is currently optional, e.g. it is not a required parameter. However, you'd error out if you didn't specify it currently. I guess I can add that to this PR 🙄

bwspenc replied:
It probably makes sense to handle that separately.  These tests already specify `solution_variables`. Maybe in a separate PR we can make that optional and then go remove `solution_variables` from all the tests.

tophmatthews replied:
Maybe I don't fully understand still...but isn't `_group_ref_resid[i]` just moving the "zero" to determine convergence? In my head, it seems that it should look more like
```
_group_resid[i] - _group_ref_resid[i] < rtol
```
I guess that is more like absolute tolerance then...

lindsayad replied:
For what you just wrote, you would have "convergence" when `_group_resid == _group_ref_resid`, e.g. when we haven't converged at all.

Relative tolerance is always based on `current_residual/initial_residual < rtol` or `current_residual/reference_residual < rtol`

tophmatthews replied:
Yeah, you're right. Perhaps the reference should be 1 if nothing is added? This becomes similar to absolute tolerance check, but still allows each variable to be evaluated separately.

tophmatthews replied:
It seems to me, in the absence of any `extra_vector_tags` for a given variable, setting `_group_ref_resid=1` will allow the 1) use of `type=ReferenceResidualProblem` and convergence to the relative tolerance, 2) comparison of the variable to its own norm (as opposed to the absolute tolerance check).

Otherwise, for thermal contact, the Lagrange multiplier will always prevent rel_tol convergence if using `ReferenceResidualProblem`, right?

lindsayad replied:
you could have a scenario where the overall L2 norm is greater than the absolute tolerance, but if all your solution variables other than the LM variable  are converged relatively and the LM variable is converged absolutely, then you would have a “relative converged” outcome

tophmatthews replied:
Yes...that is true. Kinda that the absolute tolerance comes in there, but totally true.

permcody commented:
I just turned on the Code Coverage target. I'm worried that by deleting all the MeshModifier tests without deleting the code means we'll fall below the 80% threshold. I'm not sure we'll want to merge that in just yet. I know you are trying to clear up the deprecation but removing the tests could be a non-starter. We need to get everything converted over.

permcody commented:
BTW - The holdup is that the BISON guys say that they simply can't convert everything over since MeshGenerators don't run on restart/recover. They store extra data in their current MeshModifiers. Our solution is to create a restartable datastore so that they can store their stuff in a standard location and retrieve it in other objects. As a start I made a `Parameters` object restartable, but I'll need to work out an interface to complete it. I need to get a sample of what the BISON guys need so I know I have the right requirements satisfied. 

Let's see what happened with the Code Coverage target...

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/403296/) on 0c5ade1 wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/14061/coverage)

permcody commented:
Yeah - Can't merge this. Drops us ~2.5% on code coverage.

lindsayad commented:
Ugh

permcody commented:
What don't you undo the drop of the MeshModifier tests for now. I know that won't turn our target back to green but that'll be trivial to tack back on once we are ready to pull the plug. We could try to delete the modifiers, maybe we could get most of them...

lindsayad commented:
Yea sounds good. Left that in a separate commit for a reason 😄

moosebuild commented:
Job [Code Coverage](https://civet.inl.gov/job/403362/) on 2ca95cf wanted to post the following:

View code coverage site [here](https://mooseframework.inl.gov/docs/PRs/14061/coverage)
